Configuration custom converters fail on primitive type arrays since Quarkus v3.18

### Describe the bug

Since Quarkus v3.18.0 and above, custom converters for primitive type arrays make Quarkus fail at startup with a `ClassNotFoundException` exception (`[B` class not found, so `byte[]` class).

Here's a simple converter to read a String value as a byte array:

```java
public class ByteArrayConfigConverter implements Converter<byte[]> {

    @Override
    public byte[] convert(final String s) {
        return s.getBytes(StandardCharsets.UTF_8);
    }
}
```

In Quarkus v3.17.8, Quarkus is starting correctly and the converter works. Since Quarkus v3.18.0 (still in latest 3.18.2), Quarkus is failing when starting up.

### Expected behavior

Quarkus should start correctly.

### Actual behavior

Quarkus startup fails with `ClassNotFoundException`.

```
ERROR [io.qua.run.Quarkus] (Quarkus Main Thread) Error running Quarkus: java.lang.ExceptionInInitializerError
	at io.quarkus.runner.ApplicationImpl.<clinit>(Unknown Source)
	at java.base/jdk.internal.misc.Unsafe.allocateInstance(Native Method)
	at java.base/java.lang.invoke.DirectMethodHandle.allocateInstance(DirectMethodHandle.java:501)
	at io.quarkus.runtime.Quarkus.run(Quarkus.java:76)
	at io.quarkus.runtime.Quarkus.run(Quarkus.java:48)
	at io.quarkus.runtime.Quarkus.run(Quarkus.java:137)
	at io.quarkus.runner.GeneratedMain.main(Unknown Source)
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
	at io.quarkus.runner.bootstrap.StartupActionImpl$1.run(StartupActionImpl.java:116)
	at java.base/java.lang.Thread.run(Thread.java:1583)
Caused by: java.lang.RuntimeException: java.lang.ClassNotFoundException: [B
	at io.quarkus.runtime.configuration.AbstractConfigBuilder.withConverter(AbstractConfigBuilder.java:30)
	at io.quarkus.runtime.generated.StaticInitConfigCustomizer.configBuilder(Unknown Source)
	at io.smallrye.config.SmallRyeConfigBuilder.build(SmallRyeConfigBuilder.java:770)
	at io.quarkus.runtime.generated.Config.<clinit>(Unknown Source)
	... 11 more
Caused by: java.lang.ClassNotFoundException: [B
	at java.base/jdk.internal.loader.BuiltinClassLoader.loadClass(BuiltinClassLoader.java:641)
	at java.base/jdk.internal.loader.ClassLoaders$AppClassLoader.loadClass(ClassLoaders.java:188)
	at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:526)
	at io.quarkus.bootstrap.classloading.QuarkusClassLoader.loadClass(QuarkusClassLoader.java:569)
	at io.quarkus.bootstrap.classloading.QuarkusClassLoader.loadClass(QuarkusClassLoader.java:517)
	at io.quarkus.bootstrap.classloading.QuarkusClassLoader.loadClass(QuarkusClassLoader.java:569)
	at io.quarkus.bootstrap.classloading.QuarkusClassLoader.loadClass(QuarkusClassLoader.java:517)
	at io.quarkus.runtime.configuration.AbstractConfigBuilder.withConverter(AbstractConfigBuilder.java:28)
	... 14 more
```

### How to Reproduce?

Reproducer: https://github.com/jdussouillez/quarkus-custom-converter

1. Run `./mvnw quarkus:dev` => Failure
2. Run `./mvnw quarkus:dev -Dquarkus.platform.version=3.17.8` => Works fine

### Output of `uname -a` or `ver`

_No response_

### Output of `java -version`

openjdk version "21.0.5" 2024-10-15 LTS

### Quarkus version or git rev

v3.18.0 ... v3.18.2

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

Apache Maven 3.9.9

### Additional information

I guess it's a regression because it's not mention in 3.18 migration guide. And it's not mentionned in [Microprofile config documentation](https://github.com/microprofile/microprofile-config/blob/main/spec/src/main/asciidoc/converters.asciidoc#adding-custom-converters).

A **workaround** for v3.18 is to use a record to wrap the `byte[]` field.

```java
public record KeyConfig(byte[] key) {
}

// ...

public class KeyConfigConverter implements Converter<KeyConfig> {

    @Override
    public KeyConfig convert(final String s) {
        return new KeyConfig(s.getBytes(StandardCharsets.UTF_8));
    }
}

// ...

@ConfigProperty(name = "myapp.key.value")
protected KeyConfig keyConfig;
```

--- a/core/runtime/src/main/java/io/quarkus/runtime/configuration/AbstractConfigBuilder.java
+++ b/core/runtime/src/main/java/io/quarkus/runtime/configuration/AbstractConfigBuilder.java
@@ -25,7 +25,7 @@ public abstract class AbstractConfigBuilder implements SmallRyeConfigBuilderCust
     protected static <T> void withConverter(SmallRyeConfigBuilder builder, String type, int priority, Converter<T> converter) {
         try {
             // To support converters that are not public
-            builder.withConverter((Class<T>) builder.getClassLoader().loadClass(type), priority, converter);
+            builder.withConverter((Class<T>) Class.forName(type, false, builder.getClassLoader()), priority, converter);
         } catch (ClassNotFoundException e) {
             throw new RuntimeException(e);
         }

--- a/integration-tests/test-extension/extension/deployment/src/test/java/io/quarkus/config/ByteArrayConverterTest.java
+++ b/integration-tests/test-extension/extension/deployment/src/test/java/io/quarkus/config/ByteArrayConverterTest.java
@@ -1,0 +1,28 @@
+package io.quarkus.config;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+
+import java.nio.charset.StandardCharsets;
+
+import org.eclipse.microprofile.config.inject.ConfigProperty;
+import org.jboss.shrinkwrap.api.ShrinkWrap;
+import org.jboss.shrinkwrap.api.spec.JavaArchive;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.RegisterExtension;
+
+import io.quarkus.test.QuarkusUnitTest;
+
+public class ByteArrayConverterTest {
+    @RegisterExtension
+    static final QuarkusUnitTest TEST = new QuarkusUnitTest()
+            .setArchiveProducer(() -> ShrinkWrap.create(JavaArchive.class))
+            .overrideConfigKey("someapp.bytearray", "XyZ*123");
+
+    @ConfigProperty(name = "someapp.bytearray")
+    protected byte[] bytearray;
+
+    @Test
+    void buildTimeConfigBuilder() {
+        assertEquals("XyZ*123", new String(bytearray, StandardCharsets.UTF_8));
+    }
+}

--- a/integration-tests/test-extension/extension/runtime/src/main/java/io/quarkus/extest/runtime/config/ByteArrayConfigConverter.java
+++ b/integration-tests/test-extension/extension/runtime/src/main/java/io/quarkus/extest/runtime/config/ByteArrayConfigConverter.java
@@ -1,0 +1,13 @@
+package io.quarkus.extest.runtime.config;
+
+import java.nio.charset.StandardCharsets;
+
+import org.eclipse.microprofile.config.spi.Converter;
+
+public class ByteArrayConfigConverter implements Converter<byte[]> {
+
+    @Override
+    public byte[] convert(final String s) {
+        return s.getBytes(StandardCharsets.UTF_8);
+    }
+}

--- a/integration-tests/test-extension/extension/runtime/src/main/resources/META-INF/services/org.eclipse.microprofile.config.spi.Converter
+++ b/integration-tests/test-extension/extension/runtime/src/main/resources/META-INF/services/org.eclipse.microprofile.config.spi.Converter
@@ -1,0 +1,1 @@
+io.quarkus.extest.runtime.config.ByteArrayConfigConverter

